src/datanode/src/instance/flight.rs

```diff
@@ -132,56 +135,60 @@ impl FlightService for Instance {
+        Ok(match query {
             Query::Sql(sql) => {
                 let stmt = self
                     .query_engine
                     .sql_to_statement(&sql)
-                    .context(error::ExecuteSqlSnafu)?;
+                    .context(ExecuteSqlSnafu)?;
                 self.execute_stmt(stmt, QueryContext::arc()).await?
             }
             Query::LogicalPlan(plan) => self.execute_logical(plan).await?,
-        };
-        Ok(match output {
-            Output::Stream(stream) => {
-                let stream = FlightRecordBatchStream::new(stream);
-                Box::pin(stream) as _
-            }
-            Output::RecordBatches(x) => {
-                let stream = FlightRecordBatchStream::new(x.as_stream());
-                Box::pin(stream) as _
-            }
-            Output::AffectedRows(rows) => {
-                let stream = async_stream::stream! {
-                    let ext_data = FlightDataExt {
-                        affected_rows: rows as _,
-                    }.encode_to_vec();
-                    yield Ok(FlightData::new(None, IpcMessage(build_none_flight_msg()), vec![], ext_data))
-                };
-                Box::pin(stream) as _
-            }
         })
     }
-}
 
-fn build_none_flight_msg() -> Vec<u8> {
-    let mut builder = FlatBufferBuilder::new();
+    pub async fn handle_insert(&self, request: InsertRequest) -> Result<Output> {
+        let table_name = &request.table_name.clone();
+        let table = self
+            .catalog_manager
+            .table(DEFAULT_CATALOG_NAME, &request.schema_name, table_name)
+            .context(CatalogSnafu)?
+            .context(TableNotFoundSnafu { table_name })?;
 
-    let mut message = arrow::ipc::MessageBuilder::new(&mut builder);
-    message.add_version(arrow::ipc::MetadataVersion::V5);
-    message.add_header_type(arrow::ipc::MessageHeader::NONE);
-    message.add_bodyLength(0);
+        let request = common_grpc_expr::insert::to_table_insert_request(request, table.schema())
+            .context(InsertDataSnafu)?;
 
-    let data = message.finish();
-    builder.finish(data, None);
+        let affected_rows = table
+            .insert(request)
+            .await
+            .context(InsertSnafu { table_name })?;
+        Ok(Output::AffectedRows(affected_rows))
+    }
+}
 
-    builder.finished_data().to_vec()
+fn to_flight_data_stream(output: Output) -> TonicStream<FlightData> {
+    match output {
+        Output::Stream(stream) => {
+            let stream = FlightRecordBatchStream::new(stream);
+            Box::pin(stream) as _
+        }
+        Output::RecordBatches(x) => {
+            let stream = FlightRecordBatchStream::new(x.as_stream());
+            Box::pin(stream) as _
+        }
+        Output::AffectedRows(rows) => {
+            let stream = tokio_stream::once(Ok(
+                FlightEncoder::default().encode(FlightMessage::AffectedRows(rows))
```

Comment:
Can we use a singleton of `FlightEncoder` instead of instantiating a default `FlightEncoder` every time, or make `FlightEncoder` a static method that takes default IpcWriteOptions instead of an instance method.

Reply:
I think the default initialization of encoder is negligible. It's only field IpcWriteOptions is very lightweight; and encoder's init is called only once because it's a once stream.